### ISSUE TITLE
Validate ID of vlan filter action + condition

### ISSUE THREAD

**Issue by julsemaan** [inverse-inc/packetfence#1071]
Putting a dash `-` in the condition or action ID creates a compilation error during configreload.

It should die with a more meaningful error

**Comment by jrouzierinverse** [comment]
I did not get the exact error you did. Can you provide an example conf/vlan_filters.conf

**Comment by jrouzierinverse** [comment]
I am able to get an error with a misconfigured condition

```
[category-6]
operator = is
filter = node_info.category
```

Note no value attribute.
I just want to make sure we are talking the same thing.

Error:

```
Attribute (value) does not pass the type constraint because: Validation failed for 'Str' with value undef at /usr/lib64/perl5/vendor_perl/Moose/Meta/Attribute.pm line 1279
        Moose::Meta::Attribute::verify_against_type_constraint('Moose::Meta::Attribute=HASH(0x7928168)', undef, 'instance', 'pf::condition::equals=HASH(0xaf1fa70)') called at /usr/lib64/perl5/vendor_perl/Moose/Meta/Attribute.pm line 1266
        Moose::Meta::Attribute::_coerce_and_verify('Moose::Meta::Attribute=HASH(0x7928168)', undef, 'pf::condition::equals=HASH(0xaf1fa70)') called at /usr/lib64/perl5/vendor_perl/Moose/Meta/Attribute.pm line 536
        Moose::Meta::Attribute::initialize_instance_slot('Moose::Meta::Attribute=HASH(0x7928168)', 'Moose::Meta::Instance=HASH(0x792cd78)', 'pf::condition::equals=HASH(0xaf1fa70)', 'HASH(0xaf581a0)') called at /usr/lib64/perl5/vendor_perl/Class/MOP/Class.pm line 525
        Class::MOP::Class::_construct_instance('Moose::Meta::Class=HASH(0x7921ff8)', 'HASH(0xaf581a0)') called at /usr/lib64/perl5/vendor_perl/Class/MOP/Class.pm line 498
        Class::MOP::Class::new_object('Moose::Meta::Class=HASH(0x7921ff8)', 'HASH(0xaf581a0)') called at /usr/lib64/perl5/vendor_perl/Moose/Meta/Class.pm line 284
        Moose::Meta::Class::new_object('Moose::Meta::Class=HASH(0x7921ff8)', 'HASH(0xaf581a0)') called at /usr/lib64/perl5/vendor_perl/Moose/Object.pm line 28
        Moose::Object::new('pf::condition::equals', 'HASH(0xaf61028)') called at /usr/local/pf/lib/pf/factory/condition/access_filter.pm line 79
        pf::factory::condition::access_filter::_build_sub_condition('HASH(0x7f4d21c104e8)') called at /usr/local/pf/lib/pf/factory/condition/access_filter.pm line 57
        pf::factory::condition::access_filter::instantiate('pf::factory::condition::access_filter', 'HASH(0x7f4d21c104e8)') called at /usr/local/pf/lib/pfconfig/namespaces/FilterEngine/AccessScopes.pm line 51
        pfconfig::namespaces::FilterEngine::AccessScopes::build('pfconfig::namespaces::FilterEngine::VlanScopes=HASH(0x7f4d21c...') called at /usr/local/pf/lib/pfconfig/manager.pm line 61
        pfconfig::manager::config_builder('pfconfig::manager=HASH(0xaa3edd8)', 'FilterEngine::VlanScopes') called at /usr/local/pf/lib/pfconfig/manager.pm line 284
        pfconfig::manager::cache_resource('pfconfig::manager=HASH(0xaa3edd8)', 'FilterEngine::VlanScopes') called at /usr/local/pf/lib/pfconfig/manager.pm line 362
        pfconfig::manager::expire('pfconfig::manager=HASH(0xaa3edd8)', 'FilterEngine::VlanScopes', undef) called at /usr/local/pf/lib/pfconfig/manager.pm line 370
        pfconfig::manager::expire('pfconfig::manager=HASH(0xaa3edd8)', 'config::VlanFilters') called at /usr/local/pf/lib/pfconfig/manager.pm line 478
        pfconfig::manager::expire_all('pfconfig::manager=HASH(0xaa3edd8)') called at /usr/local/pf/lib/pf/config.pm line 893
        pf::config::configreload(undef) called at /usr/local/pf/lib/pf/cmd/pf/configreload.pm line 45
        pf::cmd::pf::configreload::configreload('pf::cmd::pf::configreload=HASH(0x1b69cd8)') called at /usr/local/pf/lib/pf/cmd/pf/configreload.pm line 33
        pf::cmd::pf::configreload::action_soft('pf::cmd::pf::configreload=HASH(0x1b69cd8)') called at /usr/local/pf/lib/pf/base/cmd/action_cmd.pm line 59
        pf::base::cmd::action_cmd::_run('pf::cmd::pf::configreload=HASH(0x1b69cd8)') called at /usr/local/pf/lib/pf/cmd.pm line 27
        pf::cmd::run('pf::cmd::pf::configreload=HASH(0x1b69cd8)') called at /usr/local/pf/lib/pf/cmd/subcmd.pm line 30
        pf::cmd::subcmd::_run('pf::cmd::pf=HASH(0x1b03f28)') called at /usr/local/pf/lib/pf/cmd.pm line 27
        pf::cmd::run('pf::cmd::pf=HASH(0x1b03f28)') called at /usr/local/pf/bin/pfcmd.pl line 24
```

**Comment by julsemaan** [comment]
I think you may have put the finger on the problem.

I don't have access to the setup now but mine had an action that had no role

e.g. : 

```
[condition-with-dash]
filter = ssid
operator = is
value = BYOZammit

[action-jackson:condition-with-dash]
scope= NormalVlan
action = some_api_action
action_params = mac = $mac
```

That made the action being treated as a condition and missing the filter attribute

**Comment by julsemaan** [comment]
Note that removing the dashes made both the action and condition work (obviously without returning the role in the action)

**Comment by julsemaan** [comment]
Just confirmed the config above replicates the problem.

It tries to create a condition from the action

Here is the error during configreload :

```
Use of uninitialized value $filter in string eq at /usr/local/pf/lib/pf/factory/condition/access_filter.pm line 49.
Use of uninitialized value in hash element at /usr/local/pf/lib/pf/factory/condition/access_filter.pm line 77.
Use of uninitialized value in string eq at /usr/local/pf/lib/pf/factory/condition/access_filter.pm line 78.
Use of uninitialized value $filter in split at /usr/local/pf/lib/pf/factory/condition/access_filter.pm line 58.
Attribute (key) does not pass the type constraint because: Validation failed for 'Str' with value undef at /usr/lib64/perl5/vendor_perl/Moose/Meta/Attribute.pm line 1279
        Moose::Meta::Attribute::verify_against_type_constraint('Moose::Meta::Attribute=HASH(0x66107f0)', undef, 'instance', 'pf::condition::key=HASH(0x7f4f02cf9330)') called at /usr/lib64/perl5/vendor_perl/Moose/Meta/Attribute.pm line 1266
        Moose::Meta::Attribute::_coerce_and_verify('Moose::Meta::Attribute=HASH(0x66107f0)', undef, 'pf::condition::key=HASH(0x7f4f02cf9330)') called at /usr/lib64/perl5/vendor_perl/Moose/Meta/Attribute.pm line 536
        Moose::Meta::Attribute::initialize_instance_slot('Moose::Meta::Attribute=HASH(0x66107f0)', 'Moose::Meta::Instance=HASH(0x6613b38)', 'pf::condition::key=HASH(0x7f4f02cf9330)', 'HASH(0x7f4f02ce7ee8)') called at /usr/lib64/perl5/vendor_perl/Class/MOP/Class.pm line 525
        Class::MOP::Class::_construct_instance('Moose::Meta::Class=HASH(0x6607ea8)', 'HASH(0x7f4f02ce7ee8)') called at /usr/lib64/perl5/vendor_perl/Class/MOP/Class.pm line 498
        Class::MOP::Class::new_object('Moose::Meta::Class=HASH(0x6607ea8)', 'HASH(0x7f4f02ce7ee8)') called at /usr/lib64/perl5/vendor_perl/Moose/Meta/Class.pm line 284
        Moose::Meta::Class::new_object('Moose::Meta::Class=HASH(0x6607ea8)', 'HASH(0x7f4f02ce7ee8)') called at /usr/lib64/perl5/vendor_perl/Moose/Object.pm line 28
        Moose::Object::new('pf::condition::key', 'HASH(0x7f4f02d08cf0)') called at /usr/local/pf/lib/pf/factory/condition/access_filter.pm line 64
        pf::factory::condition::access_filter::_build_parent_condition(undef) called at /usr/local/pf/lib/pf/factory/condition/access_filter.pm line 58
        pf::factory::condition::access_filter::instantiate('pf::factory::condition::access_filter', 'HASH(0x7f4f02d1e8d8)') called at /usr/local/pf/lib/pfconfig/namespaces/FilterEngine/AccessScopes.pm line 51
        pfconfig::namespaces::FilterEngine::AccessScopes::build('pfconfig::namespaces::FilterEngine::VlanScopes=HASH(0x7f4f02c...') called at /usr/local/pf/lib/pfconfig/manager.pm line 62
        pfconfig::manager::config_builder('pfconfig::manager=HASH(0x87fda58)', 'FilterEngine::VlanScopes') called at /usr/local/pf/lib/pfconfig/manager.pm line 277
        pfconfig::manager::cache_resource('pfconfig::manager=HASH(0x87fda58)', 'FilterEngine::VlanScopes') called at /usr/local/pf/lib/pfconfig/manager.pm line 355
        pfconfig::manager::expire('pfconfig::manager=HASH(0x87fda58)', 'FilterEngine::VlanScopes', undef) called at /usr/local/pf/lib/pfconfig/manager.pm line 363
        pfconfig::manager::expire('pfconfig::manager=HASH(0x87fda58)', 'config::VlanFilters') called at /usr/local/pf/lib/pfconfig/manager.pm line 471
        pfconfig::manager::expire_all('pfconfig::manager=HASH(0x87fda58)') called at /usr/local/pf/lib/pf/config.pm line 892
        pf::config::configreload(1) called at /usr/local/pf/lib/pf/cmd/pf/configreload.pm line 45
        pf::cmd::pf::configreload::configreload('pf::cmd::pf::configreload=HASH(0x81bc78)', 1) called at /usr/local/pf/lib/pf/cmd/pf/configreload.pm line 38
        pf::cmd::pf::configreload::action_hard('pf::cmd::pf::configreload=HASH(0x81bc78)') called at /usr/local/pf/lib/pf/base/cmd/action_cmd.pm line 59
        pf::base::cmd::action_cmd::_run('pf::cmd::pf::configreload=HASH(0x81bc78)') called at /usr/local/pf/lib/pf/cmd.pm line 27
        pf::cmd::run('pf::cmd::pf::configreload=HASH(0x81bc78)') called at /usr/local/pf/lib/pf/cmd/subcmd.pm line 30
        pf::cmd::subcmd::_run('pf::cmd::pf=HASH(0x7b5f58)') called at /usr/local/pf/lib/pf/cmd.pm line 27
        pf::cmd::run('pf::cmd::pf=HASH(0x7b5f58)') called at /usr/local/pf/bin/pfcmd.pl line 24
```

**Comment by jrouzierinverse** [comment]
There are two approaches to this 
- Warn the user of the error and ignore the condition and/or rule but continue processing the rest of the rules.
- Or a full stop of processing the all the configuration.

My preferences of the first one since it will not stop the processing of configuration files on a configreload

**Comment by julsemaan** [comment]
You are right so printing a warning to stderror in configreload

And putting a warning (or fatal) in checkup would be perfect

**Comment by julsemaan** [comment]
Fixed in ?

**Comment by jrouzierinverse** [comment]
In devel

**Comment by jrouzierinverse** [comment]
Fixed in commits 134fd6fa64b0aae4f4f3ddcb4baed211ac0fda25 784b3aac49e91856e109b142f64dcf650b5b71e6 b6a15f65e149d040e5636800815b86a2a7045de5
